disperser/encoder/server_v2.go

```diff
@@ -0,0 +1,244 @@
+	return &EncoderServerV2{
+		config:      config,
+		blobStore:   blobStore,
+		chunkWriter: chunkWriter,
+		logger:      logger.With("component", "EncoderServer"),
+		prover:      prover,
+		metrics:     metrics,
+
+		runningRequests: make(chan struct{}, config.MaxConcurrentRequests),
+		requestPool:     make(chan struct{}, config.RequestPoolSize),
+	}
+}
+
+func (s *EncoderServerV2) Start() error {
+	// Serve grpc requests
+	addr := fmt.Sprintf("%s:%s", disperser.Localhost, s.config.GrpcPort)
+	listener, err := net.Listen("tcp", addr)
+	if err != nil {
+		log.Fatalf("Could not start tcp listener: %v", err)
+	}
+
+	gs := grpc.NewServer()
+	reflection.Register(gs)
+	pb.RegisterEncoderServer(gs, s)
+
+	// Register Server for Health Checks
+	name := pb.Encoder_ServiceDesc.ServiceName
+	healthcheck.RegisterHealthServer(name, gs)
+
+	s.close = func() {
+		err := listener.Close()
+		if err != nil {
+			log.Printf("failed to close listener: %v", err)
+		}
+		gs.GracefulStop()
+	}
+
+	s.logger.Info("port", s.config.GrpcPort, "address", listener.Addr().String(), "GRPC Listening")
+	return gs.Serve(listener)
+}
+
+func (s *EncoderServerV2) EncodeBlobToChunkStore(ctx context.Context, req *pb.EncodeBlobToChunkStoreRequest) (*pb.EncodeBlobToChunkStoreReply, error) {
+	totalStart := time.Now()
+	defer func() {
+		s.metrics.ObserveLatency("total", time.Since(totalStart))
+	}()
+
+	// Rate limit
+	select {
+	case s.requestPool <- struct{}{}:
+	default:
+		// TODO: Now that we no longer pass the data directly, should we pass in blob size as part of the request?
+		s.metrics.IncrementRateLimitedBlobRequestNum(1)
+		s.logger.Warn("rate limiting as request pool is full", "requestPoolSize", s.config.RequestPoolSize, "maxConcurrentRequests", s.config.MaxConcurrentRequests)
+		return nil, errors.New("too many requests")
+	}
+
+	// Limit the number of concurrent requests
+	s.runningRequests <- struct{}{}
+	defer s.popRequest()
+	if ctx.Err() != nil {
+		s.metrics.IncrementCanceledBlobRequestNum(1)
+		return nil, ctx.Err()
+	}
+
+	s.metrics.ObserveLatency("queuing", time.Since(totalStart))
+	reply, err := s.handleEncodingToChunkStore(ctx, req)
+	if err != nil {
+		s.metrics.IncrementFailedBlobRequestNum(1)
+	} else {
+		s.metrics.IncrementSuccessfulBlobRequestNum(1)
+	}
+
+	return reply, err
+}
+
+func (s *EncoderServerV2) handleEncodingToChunkStore(ctx context.Context, req *pb.EncodeBlobToChunkStoreRequest) (*pb.EncodeBlobToChunkStoreReply, error) {
+	// Validate request first
+	blobKey, encodingParams, err := s.validateAndParseRequest(req)
+	if err != nil {
+		return nil, err
+	}
+
+	s.logger.Info("Preparing to encode", "blobKey", blobKey, "encodingParams", encodingParams)
+
+	// Fetch blob data
+	fetchStart := time.Now()
+	data, err := s.blobStore.GetBlob(ctx, blobKey)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get blob from blob store: %w", err)
```

Comment:
Some errors like those in `validateAndParseRequest` are returning the canonical codes, can this and other locations be made that way as well?